File: dynamic-year.lua
local stringify = pandoc.utils.stringify

local function to_text(value)
  if type(value) == "string" then
    return value
  end

  return stringify(value)
end

local function fail(message)
  io.stderr:write("\27[31;1mERROR:\27[0m " .. message .. "\n")
  os.exit(1)
end

local function meta_integer(meta, name)
  local value = meta[name]
  if value == nil then
    return nil
  end

  local number = tonumber(to_text(value))
  if number == nil or math.floor(number) ~= number then
    fail("La metadonnee `" .. name .. "` doit etre un entier.")
  end

  return number
end

local function read_integer_arg(args, index, default, shortcode)
  local raw_value = quarto.shortcode.read_arg(args, index)
  if raw_value == nil then
    return default
  end

  local value = tonumber(to_text(raw_value))
  if value == nil or math.floor(value) ~= value then
    fail("Le shortcode `{{< " .. shortcode .. " >}}` exige un decalage entier.")
  end

  return value
end

local function base_year(meta)
  local value = meta_integer(meta, "base-year")
  if value == nil then
    fail("Les shortcodes `dynamic-year` et `dynamic-date` exigent la metadonnee `base-year` dans `_quarto.yml` ou dans le document.")
  end

  return value
end

local function is_leap_year(year)
  return year % 4 == 0 and (year % 100 ~= 0 or year % 400 == 0)
end

local function days_in_month(year, month)
  if month == 2 then
    return is_leap_year(year) and 29 or 28
  end

  if month == 4 or month == 6 or month == 9 or month == 11 then
    return 30
  end

  return 31
end

local function next_month_start(year, month)
  if month == 12 then
    return year + 1, 1, 1
  end

  return year, month + 1, 1
end

local function format_date(year, month, day)
  return string.format("%04d-%02d-%02d", year, month, day)
end

local function invalid_mode(kwargs, raw_args)
  local raw_mode = nil
  if kwargs ~= nil then
    raw_mode = kwargs["invalid"]
  end
  local mode = raw_mode ~= nil and to_text(raw_mode) or "error"
  if mode == "" and raw_args ~= nil then
    if type(raw_args) == "table" and raw_args[2] ~= nil then
      mode = to_text(raw_args[2])
    end
  end
  if mode == "" and raw_args ~= nil then
    local raw_text = to_text(raw_args)
    mode = raw_text:match("invalid%s*=%s*[\"']([^\"']+)[\"']") or raw_text:match("invalid%s*=%s*([^%s]+)") or mode
  end
  if mode == "" then
    mode = "error"
  end
  mode = mode:gsub("^['\"]", ""):gsub("['\"]$", "")
  if mode ~= "error" and mode ~= "previous-month-end" and mode ~= "next-month-start" then
    fail("Le parametre `invalid` doit valoir `error`, `previous-month-end` ou `next-month-start`.")
  end

  return mode
end

local function dynamic_date(template, year, mode)
  local month_text, day_text = template:match("^YYYY%-(%d%d)%-(%d%d)$")
  if month_text == nil then
    fail("Le shortcode `{{< dynamic-date >}}` exige une date sous la forme `YYYY-MM-DD`, par exemple `YYYY-10-15`.")
  end

  local month = tonumber(month_text)
  local day = tonumber(day_text)
  if month < 1 or month > 12 then
    fail("Le shortcode `{{< dynamic-date >}}` contient un mois invalide.")
  end

  local max_day = days_in_month(year, month)
  if day >= 1 and day <= max_day then
    return format_date(year, month, day)
  end

  if day < 1 then
    fail("Le shortcode `{{< dynamic-date >}}` contient un jour invalide.")
  end

  if mode == "previous-month-end" then
    return format_date(year, month, max_day)
  end

  if mode == "next-month-start" then
    local next_year, next_month, next_day = next_month_start(year, month)
    return format_date(next_year, next_month, next_day)
  end

  fail("La date `" .. format_date(year, month, day) .. "` est invalide. Utilisez `invalid=\"previous-month-end\"` ou `invalid=\"next-month-start\"` pour la corriger automatiquement.")
end

return {
  ["dynamic-year"] = function(args, _kwargs, meta, _raw_args, _context)
    local offset = read_integer_arg(args, 1, 0, "dynamic-year x")
    return pandoc.Str(tostring(base_year(meta) + offset))
  end,
  ["dynamic-date"] = function(args, kwargs, meta, raw_args, _context)
    local raw_template = quarto.shortcode.read_arg(args, 1)
    if raw_template == nil then
      fail("Le shortcode `{{< dynamic-date >}}` exige une date sous la forme `YYYY-MM-DD`, par exemple `YYYY-10-15`.")
    end

    local offset = read_integer_arg(args, 2, 0, "dynamic-date \"YYYY-MM-DD\" x")
    local year = base_year(meta) + offset
    local mode = invalid_mode(kwargs, raw_args)

    return pandoc.Str(dynamic_date(to_text(raw_template), year, mode))
  end
}

File: dynamic-year.R
dynamic_base_year <- function(path = NULL) {
  value <- NULL

  if (is.null(path)) {
    value <- read_base_year_from_current_document()
    if (is.null(value)) {
      value <- read_base_year_from_project()
    }
  } else {
    value <- read_base_year_from_file(path)
  }

  if (is.null(value)) {
    stop("`base-year` est requis dans le front matter du document ou dans `_quarto.yml`.", call. = FALSE)
  }

  validate_integer(value, "`base-year`")
}

dynamic_year <- function(offset = 0, base_year = dynamic_base_year()) {
  validate_integer(base_year, "`base_year`") + validate_integer(offset, "`offset`")
}

dynamic_date <- function(x, offset = 0, invalid = c("error", "previous-month-end", "next-month-start"),
                         base_year = dynamic_base_year()) {
  invalid <- match.arg(invalid)
  target_year <- dynamic_year(offset = offset, base_year = base_year)

  if (inherits(x, "Date")) {
    value <- vapply(x, dynamic_date_one, character(1), target_year, invalid, "Date")
    return(as.Date(unname(value)))
  }

  if (is.numeric(x)) {
    year <- rep.int(target_year, length(x))
    year[is.na(x)] <- NA_integer_
    return(year)
  }

  if (is.character(x)) {
    value <- vapply(x, dynamic_date_one, character(1), target_year, invalid, "character")
    return(unname(value))
  }

  stop("`x` doit etre un objet Date, une chaine `YYYY-MM-DD`, une chaine `YYYY` ou une annee numerique.", call. = FALSE)
}

dynamic_date_one <- function(x, target_year, invalid, output_type) {
  if (is.na(x)) {
    return(NA_character_)
  }

  text <- if (output_type == "Date") format(x, "%Y-%m-%d") else x

  if (grepl("^(YYYY|[0-9]{4})$", text)) {
    return(sprintf("%04d", target_year))
  }

  parts <- regexec("^(YYYY|[0-9]{4})-([0-9]{2})-([0-9]{2})$", text)
  match <- regmatches(text, parts)[[1]]
  if (length(match) == 0) {
    stop("`x` doit utiliser le format `YYYY-MM-DD`, `yyyy-mm-dd` ou `yyyy`.", call. = FALSE)
  }

  month <- as.integer(match[[3]])
  day <- as.integer(match[[4]])
  build_dynamic_date(target_year, month, day, invalid)
}

build_dynamic_date <- function(year, month, day, invalid) {
  if (month < 1 || month > 12) {
    stop("La date dynamique contient un mois invalide.", call. = FALSE)
  }

  max_day <- days_in_month(year, month)
  if (day >= 1 && day <= max_day) {
    return(sprintf("%04d-%02d-%02d", year, month, day))
  }

  if (day < 1) {
    stop("La date dynamique contient un jour invalide.", call. = FALSE)
  }

  if (invalid == "previous-month-end") {
    return(sprintf("%04d-%02d-%02d", year, month, max_day))
  }

  if (invalid == "next-month-start") {
    if (month == 12) {
      return(sprintf("%04d-01-01", year + 1))
    }
    return(sprintf("%04d-%02d-01", year, month + 1))
  }

  stop(
    sprintf(
      "La date `%04d-%02d-%02d` est invalide. Utilisez invalid = \"previous-month-end\" ou invalid = \"next-month-start\" pour la corriger automatiquement.",
      year, month, day
    ),
    call. = FALSE
  )
}

days_in_month <- function(year, month) {
  if (month == 2) {
    return(if (is_leap_year(year)) 29L else 28L)
  }
  if (month %in% c(4L, 6L, 9L, 11L)) {
    return(30L)
  }
  31L
}

is_leap_year <- function(year) {
  year %% 4 == 0 && (year %% 100 != 0 || year %% 400 == 0)
}

validate_integer <- function(value, name) {
  number <- suppressWarnings(as.numeric(value))
  if (length(number) != 1 || is.na(number) || number != floor(number)) {
    stop(name, " doit etre un entier.", call. = FALSE)
  }
  as.integer(number)
}

read_base_year_from_current_document <- function() {
  candidates <- character()

  if (requireNamespace("knitr", quietly = TRUE)) {
    input <- knitr::current_input()
    if (!is.null(input) && nzchar(input)) {
      candidates <- c(candidates, sub("\\.rmarkdown$", ".qmd", input))
    }
  }

  candidates <- c(candidates, "index.qmd")
  candidates <- unique(file.path(getwd(), candidates))
  candidates <- candidates[file.exists(candidates)]

  for (candidate in candidates) {
    value <- read_base_year_from_file(candidate)
    if (!is.null(value)) {
      return(value)
    }
  }

  NULL
}

read_base_year_from_project <- function() {
  starts <- c(Sys.getenv("QUARTO_PROJECT_DIR", unset = ""), getwd())
  starts <- unique(normalizePath(starts[nzchar(starts)], mustWork = FALSE))

  for (start in starts) {
    current <- start
    repeat {
      candidate <- file.path(current, "_quarto.yml")
      value <- read_base_year_from_file(candidate)
      if (!is.null(value)) {
        return(value)
      }

      parent <- dirname(current)
      if (identical(parent, current)) {
        break
      }
      current <- parent
    }
  }

  NULL
}

read_base_year_from_file <- function(path) {
  if (is.null(path) || !file.exists(path)) {
    return(NULL)
  }

  lines <- readLines(path, warn = FALSE)
  if (length(lines) == 0) {
    return(NULL)
  }

  if (identical(trimws(lines[[1]]), "---")) {
    end <- which(trimws(lines[-1]) == "---")[1]
    if (!is.na(end)) {
      lines <- lines[seq_len(end + 1)]
    }
  }

  match <- regmatches(lines, regexec("^\\s*base-year\\s*:\\s*['\"]?([+-]?[0-9]+)['\"]?\\s*$", lines))
  values <- vapply(match, length, integer(1))
  if (!any(values > 1)) {
    return(NULL)
  }

  match[[which(values > 1)[1]]][[2]]
}
